File: rfcs/bench-0104/harness/js/spectralnorm.js
// spectral-norm, the JavaScript leg of RFC-0104 M2.
//
// The specification is examples/spectralnorm.vyrn. Same algorithm, same
// expression order, byte-identical output.
//
//   $ node spectralnorm.js         # the census N, order 100
//   $ node spectralnorm.js 3000    # the bench N
//
// Structural differences from the Vyrn program:
//
//   * Math.sqrt is the scalar square root Vyrn reaches through F64x2.sqrt(..).lane(0).
//   * multiplyAv / multiplyAtv write into the destination in place. Vyrn's
//     ownership model makes the destination `consume`-in / owned-out; JS has no
//     such rule, so no move-in-move-out shape is written.
//   * Float64Array here, Array<Float64> there — the same flat run of doubles.

'use strict';
const fs = require('fs');

/// The census N — a 100x100 window of the matrix, the size the fixture was
/// written at.
const order = 100;

let pending = '';
function emit(s) {
    pending += s;
    if (pending.length >= 1 << 20) flush();
}
function flush() {
    if (pending.length === 0) return;
    const buf = Buffer.from(pending, 'latin1');
    pending = '';
    let off = 0;
    while (off < buf.length) {
        try {
            off += fs.writeSync(1, buf, off, buf.length - off);
        } catch (e) {
            if (e.code !== 'EAGAIN') throw e;
        }
    }
}

/// `v` at nine decimal places, the format the game prints.
function fixed9(v) {
    let a = v;
    let sign = '';
    if (a < 0.0) {
        sign = '-';
        a = 0.0 - a;
    }
    const scaled = Math.trunc(a * 1000000000.0 + 0.5);
    const whole = Math.trunc(scaled / 1000000000);
    const frac = String(scaled % 1000000000).padStart(9, '0');
    return `${sign}${whole}.${frac}`;
}

/// `A[i][j]` — the matrix is a formula, so no matrix is ever built.
function cell(i, j) {
    return 1.0 / (Math.trunc((i + j) * (i + j + 1) / 2) + i + 1);
}

/// `out = A v`.
function multiplyAv(n, v, w) {
    let i = 0;
    while (i < n) {
        let sum = 0.0;
        let j = 0;
        while (j < n) {
            sum = sum + cell(i, j) * v[j];
            j = j + 1;
        }
        w[i] = sum;
        i = i + 1;
    }
    return w;
}

/// `out = A-transpose v`, which is multiplyAv with the indices of cell swapped.
function multiplyAtv(n, v, w) {
    let i = 0;
    while (i < n) {
        let sum = 0.0;
        let j = 0;
        while (j < n) {
            sum = sum + cell(j, i) * v[j];
            j = j + 1;
        }
        w[i] = sum;
        i = i + 1;
    }
    return w;
}

/// Ten rounds of `u = A-transpose A u`, then the Rayleigh quotient's square root.
function spectralNorm(n) {
    let u = new Float64Array(n).fill(1.0);
    let v = new Float64Array(n).fill(0.0);
    let w = new Float64Array(n).fill(0.0);
    let round = 0;
    while (round < 10) {
        w = multiplyAv(n, u, w);
        v = multiplyAtv(n, w, v);
        w = multiplyAv(n, v, w);
        u = multiplyAtv(n, w, u);
        round = round + 1;
    }
    let vbv = 0.0;
    let vv = 0.0;
    let k = 0;
    while (k < n) {
        vbv = vbv + u[k] * v[k];
        vv = vv + v[k] * v[k];
        k = k + 1;
    }
    return Math.sqrt(vbv / vv);
}

const n = process.argv[2] === undefined ? order : Number(process.argv[2]);
emit(fixed9(spectralNorm(n)) + '\n');
flush();

File: rfcs/bench-0104/harness/rust/spectralnorm.rs
// spectral-norm, from the Computer Language Benchmarks Game.
//
// The Rust leg of RFC-0104 M2. It computes what `examples/spectralnorm.vyrn`
// computes, in the same expression order, and prints the same bytes.
//
//   $ rustc -C opt-level=3 -o spectralnorm.exe spectralnorm.rs
//   $ ./spectralnorm 100
//
// Safe Rust, std only. Single threaded: the Vyrn program is, and the row is a
// measurement of the numeric path and not of a thread pool.

use std::env;

/// The census N — a 100x100 window of the matrix.
const ORDER: i64 = 100;

/// `v` at nine decimal places, the format the game prints.
fn fixed9(v: f64) -> String {
    let mut a = v;
    let mut sign = "";
    if a < 0.0 {
        sign = "-";
        a = 0.0 - a;
    }
    let scaled = (a * 1000000000.0 + 0.5) as i64;
    format!("{}{}.{:09}", sign, scaled / 1000000000, scaled % 1000000000)
}

/// `A[i][j]` — the matrix is a formula, so no matrix is ever built.
fn cell(i: i64, j: i64) -> f64 {
    1.0 / (((i + j) * (i + j + 1) / 2 + i + 1) as f64)
}

/// `out = A v`.
fn multiply_av(n: i64, v: &[f64], out: &mut [f64]) {
    for i in 0..n {
        let mut sum = 0.0;
        for j in 0..n {
            sum = sum + cell(i, j) * v[j as usize];
        }
        out[i as usize] = sum;
    }
}

/// `out = A-transpose v`, which is `multiply_av` with the indices of `cell` swapped.
fn multiply_atv(n: i64, v: &[f64], out: &mut [f64]) {
    for i in 0..n {
        let mut sum = 0.0;
        for j in 0..n {
            sum = sum + cell(j, i) * v[j as usize];
        }
        out[i as usize] = sum;
    }
}

/// Ten rounds of `u = A-transpose A u`, then the Rayleigh quotient's square root.
fn spectral_norm(n: i64) -> f64 {
    let mut u = vec![1.0; n as usize];
    let mut v = vec![0.0; n as usize];
    let mut w = vec![0.0; n as usize];
    for _ in 0..10 {
        multiply_av(n, &u, &mut w);
        multiply_atv(n, &w, &mut v);
        multiply_av(n, &v, &mut w);
        multiply_atv(n, &w, &mut u);
    }
    let mut vbv = 0.0;
    let mut vv = 0.0;
    for k in 0..n as usize {
        vbv = vbv + u[k] * v[k];
        vv = vv + v[k] * v[k];
    }
    (vbv / vv).sqrt()
}

fn main() {
    let n: i64 = env::args()
        .nth(1)
        .map(|a| a.parse().expect("N must be an integer"))
        .unwrap_or(ORDER);

    println!("{}", fixed9(spectral_norm(n)));
}
